### PR TITLE
Add activity roles

### PR DESCRIPTION
**Pull request by nicksellen**
Forum thread: https://community.foodsaving.world/t/applicant-trial-pickup-proposal/575
Frontend PR: https://github.com/yunity/karrot-frontend/pull/2421

This is a continuation of https://github.com/yunity/karrot-backend/pull/1105 but it took a slightly different path.

Previously:
- activities could function just as normal
- optionally you could require a certain role to join
- and if you did that, you could optionally also allow any group members to join in "open" slots

Now:
- there are **participant types**
- each declares a name, required role, max_participants, and a description
- it must always be declared explicitly now
- there can be any number of them participant types per activity, or series
- there is a new "member" role that all group members will _always_ have (enforced by db constraint)

The same:
- there is no extra roles, or mechanism to get them (see https://github.com/yunity/karrot-backend/pull/1167 for that work)
- roles are still not customizable, that comes in the future

## TODO:

- [x] add API for updating participant roles for activity
- [x] add API for updating participant roles for activity series
- [x] handle updating activities with a series when changing participant roles
- [x] send notification if user removed from activity due to role requirement change?
- [x] rework notifications (push, reminder, emails) to be role aware
  - [x] activity_notification email (was already done)
  - [x] group_summary email (no change necessary)
  - [x] activity_reminder push (no change necessary) 
  - [x] activity_upcoming, activity_disabled, activity_enabled, activity_moved notifications (no change necessary)
  - [x] history
- [x] ~~make `is_modified` available in API (for UI)~~ I removed it again, not needed now
- [x] add tests for participant role API
- [x] deal with the TODO items in the code I left...
- [x] remove max_participants / require_role / max_open_participants from models/API/etc

### CODE REVIEW

**Comment by tiltec** [comment]
I'll have a look at fixing the tests, then I could look into making existing notifications role-aware.

**Comment by nicksellen** [comment]
I think the existing notifications should already be role aware. Although I could have missed some ..

I'm sure it could be improved too,

-------- Original Message --------

**Comment by tiltec** [comment]
Yeah, notification were alright! I just saw the unticked checkbox in the PR description, so I thought this might be a nice small task :)

**Review by tiltec — COMMENTED** [review]
Hey, great work! I tried to understand most of it, but it's not simple... anyway, I left a few comments.

I didn't find a test for the new `/check/` endpoint, maybe you could add a quick one?

I didn't find tests for modifying a series when an individual pickup was changed - from the code it looks like it's possible, but I didn't try it out. What's the intended behavior?

File: karrot/places/serializers.py [hunk 1]
```diff
@@ -184,14 +184,14 @@ class Meta:
...
 
+    @transaction.atomic()
     def save(self, **kwargs):
         self._validated_data = find_changed(self.instance, self.validated_data)
         skip_update = len(self.validated_data.keys()) == 0
         if skip_update:
             return self.instance
         return super().save(**kwargs)
 
-    @transaction.atomic()
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/places/serializers.py:194]: I moved `@transaction.atomic()` to the `save()` call instead (which in turns calls `update()`), so it's still atomic.

File: karrot/activities/serializers.py [hunk 2]
```diff
@@ -572,40 +708,275 @@ def get_instance_attr(field):
...
+                                        # find all the participants who are missing the new role, and remove them...
+                                        users_with_new_role = series.place.group.members.filter(
+                                            groupmembership__roles__contains=[role]
+                                        )
+                                        participants = ActivityParticipant.objects.filter(
+                                            participant_type=participant_type,
+                                            activity__in=activities,
+                                        ).exclude(user__in=users_with_new_role)
+                                        for participant in participants:
+                                            add_removed(participant.activity, participant.user)
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/activities/serializers.py:837]: yes, it's a bit of a beast... I could have a go at splitting it into multiple methods...

File: karrot/activities/tests/test_activities_api.py [hunk 3]
```diff
@@ -529,7 +555,162 @@ def test_export_ics_activities_group(self):
...
+            '/api/activities/{}/'.format(self.activity.id), {'participant_types': [{
+                'id': pt.id,
+                'role': 'cook',
+            }]},
+            format='json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_200_OK, response.data)
+        self.assertEqual(self.activity.participants.count(), 0)
+
+    def test_remove_participant_type(self):
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/activities/tests/test_activities_api.py:694]: there is a little secret that the fancy "removed participants" stuff is only for series changes... perhaps I can add it for that (especially as it doesn't need the fancy _check_ method, as the frontend knows if someone is signed up for an individual activity...

File: karrot/activities/tests/test_activity_series_api.py [hunk 4]
```diff
@@ -181,28 +222,126 @@ def setUp(self):
...
+                'role': 'driver',
+            }]},
+            format='json'
+        )
+        self.group.refresh_from_db()
+        self.assertEqual(
+            list(self.series.participant_types.order_by('role').values_list('role', flat=True)),
+            ['approved', 'driver', 'member'],
+        )
+        self.series.activities.upcoming()
```
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/activities/tests/test_activity_series_api.py:276]: not sure, I just removed it...

**Comment by nicksellen** [comment]
> I didn't find tests for modifying a series when an individual ~~pickup~~ _activity_ was changed - from the code it looks like it's possible, but I didn't try it out. What's the intended behavior?

yes, should update everything that wasn't changed in the individual activities, including nested participant type fields, e.g.:
- if you change the _description_ in a participant type for a activity that is part of a series...
- ... then modify the _role_ for the participant type in that series
- ... the change to _role_ will be made to the activity that was edited, and the _description_ will keep it's edited value

... I originally wasn't going to make it do such fiddly changes, and make it cruder, but went the whole way in the end.

the frontend understands this too, e.g. here the role has been modified from the series value for that particular participant type:



... this is possible because participant types are created per-activity, but still referenced back to the _series participant type_ that they came from.

yes, more tests there would be good, my energy for this PR is going down a lot though! would :heart: to merge it soon!

... still waiting on a few changes to how the wording/notifications when participants are removed is done, so not quite ready.

File: karrot/activities/tests/test_activities_api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiltec) [reply karrot/activities/tests/test_activities_api.py:694]: nice! I didn't notice that it was just for series before...

File: karrot/activities/serializers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicksellen) [reply karrot/activities/serializers.py:889]: I've left it as long methods for now... but there is still potential for extracting parts of it (esp. to reuse with the check method), but gotta get it merged now!

**Pull request merged**